pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/blocks/DataBlockBuilder.java

```diff
@@ -0,0 +1,406 @@
+    _numColumns = dataSchema.size();
+    if (_blockType == BaseDataBlock.Type.COLUMNAR) {
+      _cumulativeColumnOffsetSizeInBytes = new int[_numColumns];
+      _columnSizeInBytes = new int[_numColumns];
+      DataBlockUtils.computeColumnSizeInBytes(_dataSchema, _columnSizeInBytes);
+      int cumulativeColumnOffset = 0;
+      for (int i = 0; i < _numColumns; i++) {
+        _cumulativeColumnOffsetSizeInBytes[i] = cumulativeColumnOffset;
+        cumulativeColumnOffset += _columnSizeInBytes[i] * _numRows;
+      }
+    } else if (_blockType == BaseDataBlock.Type.ROW) {
+      _columnOffsets = new int[_numColumns];
+      _rowSizeInBytes = DataTableUtils.computeColumnOffsets(dataSchema, _columnOffsets);
+    }
+  }
+
+  public static RowDataBlock buildFromRows(List<Object[]> rows, DataSchema dataSchema)
+      throws IOException {
+    DataBlockBuilder rowBuilder = new DataBlockBuilder(dataSchema, BaseDataBlock.Type.ROW);
+    rowBuilder._numRows = rows.size();
+    for (Object[] row : rows) {
+      ByteBuffer byteBuffer = ByteBuffer.allocate(rowBuilder._rowSizeInBytes);
+      for (int i = 0; i < rowBuilder._numColumns; i++) {
+        Object value = row[i];
+        switch (rowBuilder._columnDataType[i]) {
+          // Single-value column
+          case INT:
+            byteBuffer.putInt(((Number) value).intValue());
+            break;
+          case LONG:
+            byteBuffer.putLong(((Number) value).longValue());
+            break;
+          case FLOAT:
+            byteBuffer.putFloat(((Number) value).floatValue());
+            break;
+          case DOUBLE:
+            byteBuffer.putDouble(((Number) value).doubleValue());
+            break;
+          case BIG_DECIMAL:
+            setColumn(rowBuilder, byteBuffer, (BigDecimal) value);
+            break;
+          case STRING:
+            setColumn(rowBuilder, byteBuffer, i, (String) value);
+            break;
+          case BYTES:
+            setColumn(rowBuilder, byteBuffer, (ByteArray) value);
+            break;
+          case OBJECT:
+            setColumn(rowBuilder, byteBuffer, value);
+            break;
+          // Multi-value column
+          case BOOLEAN_ARRAY:
+          case INT_ARRAY:
+            setColumn(rowBuilder, byteBuffer, (int[]) value);
+            break;
+          case TIMESTAMP_ARRAY:
+          case LONG_ARRAY:
+            // LONG_ARRAY type covers INT_ARRAY and LONG_ARRAY
+            if (value instanceof int[]) {
+              int[] ints = (int[]) value;
+              int length = ints.length;
+              long[] longs = new long[length];
+              ArrayCopyUtils.copy(ints, longs, length);
+              setColumn(rowBuilder, byteBuffer, longs);
+            } else {
+              setColumn(rowBuilder, byteBuffer, (long[]) value);
+            }
+            break;
+          case FLOAT_ARRAY:
+            setColumn(rowBuilder, byteBuffer, (float[]) value);
+            break;
+          case DOUBLE_ARRAY:
```

Comment:
I guess this is dictated by the root operator in the current stage that is outputting data to be transferred. So, it knows the type and that is what we should use here in the schema when building the `DataBlock`. So, why do we need to handle int/long/float if we know the final type after processing the current stage is DOUBLE ?

Reply:
i copied the code from DataTableBuilder :-)